### ISSUE TITLE
Technical regression for 'hwmontemp_linux' in c91167f (not a bug)

### ISSUE THREAD

**Issue by lynix** [vicious-widgets/vicious#98]
Well, I don't know if this is the right place to start a discussion like this, but as the commit I'm going to comment on has been merged back in June 2019 I think it's better to open an issue instead of adding a comment directly to it.

I'm the original author of `hwmontemp_linux` and I just noticed what changes c91167f (by @McSinyx) has brought to it.

The commit message says it's deprecating `io.popen`. Fair enough, but in my eyes that commit is nothing less than a re-write of the entire widget. Nothing wrong with that in general, but there are two things that I really dislike about the new code:

1) It throws away the path caching mechanism, so the correct sysfs path is searched in each invocation
2) It depends on the `grep` binary to be installed and calls it every time via shell

Please tell me if I'm wrong and just did not understand the new code. That widget has been my first piece of Lua code and I'm still learning.

Assuming I did understand the code correctly and my two points are valid, I'd like to know if you were open for a pull request of mine that would bring back the cache and get rid of `grep`. IMHO it's quite overkill to spawn a separate process (and shell) just to get the contents of a file, when you're already running Lua.

Best regards,

lynix

**Comment by mutlusun** [comment]
Hello @lynix ,
thanks for your issue! I think everyone of us is happy about contributions to this project. So feel free to create a pull request! ;) I think it's easier to discuss any changes there and it is probably better to remove the reliance on `grep`. However, I suggest to keep `io.popen` out of the game.

Best,
mutlusun

**Comment by McSinyx** [comment]
Sorry for the late reply, but I'm also on board with any contribution that will improve the widget's performance.  Regarding grep, I think its usage is POSIX-compliant (I recall checking for every shell command during the refactoring progress) so compatibility should not be a concern, but intuitively I agree that it might be well slower than a pure-Lua solution.

**Comment by lynix** [comment]
Okay so I had a look into this today. Turns out that my original implementation wasn't pure-Lua either, spawning `ls -1` for iterating over sysfs directory entries. So I kept `grep` and just brought back the cache. This way we only spawn `grep` for cache misses, which should only happen once during normal usage.

Pull request: #99 

I'm closing this issue as discussion regarding the pull request can happen therein. Thanks!
